### PR TITLE
Move pull requests to v4

### PR DESCRIPTION
**Pull request by mknoedel**
Moved Pull Request ingestion to Github's GraphQL endpoint and made some refactors. Doing this gives us several benefits:
1. We are now able to fetch pull requests using [the GitHub search syntax](https://docs.github.com/en/github/searching-for-information-on-github/searching-on-github/searching-issues-and-pull-requests), which in turn opens the door for the rolling ingestion of Pull Requests with the `updatedOn` search term.
2. We are able to pull commits and reviews along with each PR which eliminates the amount of calls to the GitHub apis (which have extremely restrictive rate limits) by about a factor of 3, and makes commit analysis trivial. Commit analysis was the source of many ongoing customer issues so those should be fixed as well.
3. This integration is much faster now. Pull requests were by far the slowest step and this makes it 6.5x faster, from ~57 minutes to 9 minutes from my testing. This equates to a 4x overall integration speed improvement, from ~82 minutes to ~19 minutes total execution time from my testing.

Old
image

New
image

### CODE REVIEW

File: src/client/GraphQLClient/__snapshots__/client.test.ts.snap [hunk 1]
```diff
@@ -0,0 +1,1581 @@
...
+    "additions": 6,
+    "author": Object {},
+    "authorAssociation": "CONTRIBUTOR",
+    "baseRefName": "main",
+    "baseRefOid": "4f2c15b0346be5f56755bc0291f2d387bec6bd21",
+    "baseRepository": Object {
+      "id": "MDEwOlJlcG9zaXRvcnkxNzE5MTExMzk=",
+      "name": "graph-whitehat",
+    },
+    "body": "Bumps [ws](https://github.com/websockets/ws) from 5.2.2 to 5.2.3.
```
Comment (mknoedel) [inline src/client/GraphQLClient/__snapshots__/client.test.ts.snap:274]: I think we should likely limit how much of the Pull Request body we put into the rawData

File: src/client/GraphQLClient/buildGraphQL.test.ts [hunk 2]
```diff
@@ -1,6 +1,115 @@
...
+
+  const pageLimit = 1;
+  const metadataMap = resourceMetadataMap(pageLimit);
+  const gql = (queryResources: PullRequestResource[]) => {
+    return buildGraphQL(
+      metadataMap,
+      PullRequestResource.PullRequests,
+      queryResources,
+    );
+  };
```
Comment (mknoedel) [inline src/client/GraphQLClient/buildGraphQL.test.ts:18]: - [ ] Clean this up

File: src/client/GraphQLClient/buildGraphQL.ts [hunk 3]
```diff
@@ -1,16 +1,25 @@
...
   resourceMetadataMap: ResourceMap<ResourceMetadata>,
-  queryResources: OrganizationResource[]
+  parentResource: GithubResource,
+  queryResources: GithubResource[],
 ): string {
-  const variables: string[] = ['$login: String!'];
   const queries: QueryHierarchy[] = [];
+  const variables: string[] = [];
+
+  // TODO: make this not super ugly
```
Comment (mknoedel) [inline src/client/GraphQLClient/buildGraphQL.ts:16]: - [ ] Make this not super ugly

File: src/client/GraphQLClient/buildGraphQL.ts [hunk 4]
```diff
@@ -25,34 +34,42 @@ export default function buildGraphQL(
       resourceMetadata = resourceMetadataMap[parent];
     }
 
-    variables.push(resourceMetadata.graphRequestVariable);
+    if (resourceMetadata.graphRequestVariable) {
```
Comment (mknoedel) [inline src/client/GraphQLClient/buildGraphQL.ts:37]: - [ ] Also make this not super ugly

File: src/client/GraphQLClient/buildGraphQL.ts [hunk 5]
```diff
@@ -25,34 +34,42 @@ export default function buildGraphQL(
...
 
     if (resourceMetadata.children && includedChildren.length > 0) {
-      variables.push(
-        ...includedChildren.map(
-          (c: OrganizationResource) =>
-            resourceMetadataMap[c].graphRequestVariable
-        )
-      );
+      includedChildren.forEach((c: GithubResource) => {
+        if (resourceMetadataMap[c].graphRequestVariable) {
```
Comment (mknoedel) [inline src/client/GraphQLClient/buildGraphQL.ts:59]: - [ ] Make this not super ugly as well

Comment (mknoedel) [inline src/client/GraphQLClient/buildGraphQL.ts:59]: - [ ] Make this not super ugly as well

File: src/client/GraphQLClient/client.test.ts [hunk 6]
```diff
@@ -18,7 +23,7 @@ async function getAccess() {
   const config = context.instance.config;
   sanitizeConfig(config);
   //the installid in the recordings
-  config.installationId = 17214088;
+  config.installationId = 7498286;
```
Comment (mknoedel) [inline src/client/GraphQLClient/client.test.ts:26]: - [ ] Make the installationId be pulled from the environment

File: src/client/GraphQLClient/client.test.ts [hunk 7]
```diff
@@ -31,15 +36,85 @@ async function getAccess() {
...
+  });
+
+  test('pullRequest pagination only', async () => {
+    p = setupGithubRecording({
+      directory: __dirname,
+      name: 'GitHubGraphQLClient.fetchPullRequests.singlePage',
+      options: {
+        matchRequestsBy: {
+          headers: false, //must not set order:false
+        },
```
Comment (mknoedel) [inline src/client/GraphQLClient/client.test.ts:63]: I don't think this is needed

File: src/client/GraphQLClient/client.test.ts [hunk 8]
```diff
@@ -31,15 +36,85 @@ async function getAccess() {
...
 }
 
+describe('pull requests', () => {
+  let p: Recording; //p for polly
+
+  afterEach(async () => {
+    await p.stop();
+  });
+
+  test('pullRequest pagination only', async () => {
```
Comment (mknoedel) [inline src/client/GraphQLClient/client.test.ts:56]: Need one more test without asking for any extra resources.

File: src/approval/__recordings__/collectCommitsForPR-dismissedReview_1084922421/recording.har [hunk 9]
```diff
@@ -1,500 +0,0 @@
-{
```
Comment (mknoedel) [inline src/approval/__recordings__/collectCommitsForPR-dismissedReview_1084922421/recording.har:1]: These are no longer needed because we fetch commits with PRs

File: package.json [hunk 10]
```diff
@@ -37,8 +37,6 @@
     "koa": "^2.11.0",
     "koa-bodyparser": "^4.2.1",
     "koa-router": "^7.4.0",
-    "lodash.omit": "^4.5.0",
-    "lodash.uniq": "^4.5.0",
```
Comment (mknoedel) [inline package.json:41]: lodash was already installed in this project

File: src/approval/collectCommitsForPR.ts [hunk 11]
```diff
@@ -1,146 +0,0 @@
-import OrganizationAccountClient from '../client/OrganizationAccountClient';
```
Comment (mknoedel) [inline src/approval/collectCommitsForPR.ts:1]: This method is no longer necessary as we fetch commits with PRs

File: src/client.ts [hunk 12]
```diff
@@ -197,17 +181,6 @@ export class APIClient {
...
-
-    //we would prefer to use GraphQL to get collabs, but we haven't figured out how to make that work
-    //this code for future dev
-    /*
-    const collabs: OrgCollaboratorQueryResponse[] = await this.accountClient.getRepoCollaborators();
-    console.log('GraphQL approach to collabs:');
-    console.log(collabs);
-    for (const collab of collabs) {
-      console.log(collab);
-    }
```
Comment (mknoedel) [inline src/client.ts:209]: This will be done in another PR

File: src/client.ts [hunk 13]
```diff
@@ -154,29 +147,20 @@ export class APIClient {
...
-
-    const pullrequests = await this.accountClient.getPullRequestEntities(
-      account,
-      repo,
-      memberEntities,
-      memberByLoginMap,
-      logger,
+    const {
+      rateLimitConsumed,
+    } = await this.accountClient.iteratePullRequestEntities(repo, iteratee);
```
Comment (mknoedel) [inline src/client.ts:159]: Iterate instead of get in order to save on memory

File: src/client.ts [hunk 14]
```diff
@@ -233,7 +206,7 @@ export class APIClient {
     } catch (err) {
       throw new IntegrationProviderAuthenticationError({
         cause: err,
-        endpoint: `https://api.github.com/app/installations/${this.config.installation_id}/access_tokens`,
+        endpoint: `https://api.github.com/app/installations/${this.config.installationId}/access_tokens`,
```
Comment (mknoedel) [inline src/client.ts:209]: Bonus change: this was always `undefined` in the error messages.

File: src/client/GraphQLClient/buildGraphQL.ts [hunk 15]
```diff
@@ -1,16 +1,23 @@
...
 ): string {
-  const variables: string[] = ['$login: String!'];
   const queries: QueryHierarchy[] = [];
+  const variables: string[] = [];
+
+  if (resourceMetadataMap[parentResource].graphRequestVariables) {
+    variables.push(
+      ...resourceMetadataMap[parentResource].graphRequestVariables,
+    );
+  }
```
Comment (mknoedel) [inline src/client/GraphQLClient/buildGraphQL.ts:20]: Added support for specifying any amount of graph variables, not just 1.

File: src/client/GraphQLClient/buildGraphQL.ts [hunk 16]
```diff
@@ -61,13 +68,13 @@ export default function buildGraphQL(
 
   const fragmentHierarchy: QueryHierarchy[] = [
     {
-      self: resourceMetadataMap[OrganizationResource.Organization].factory,
-      children: [...queries]
+      self: resourceMetadataMap[parentResource].factory,
```
Comment (mknoedel) [inline src/client/GraphQLClient/buildGraphQL.ts:71]: Pull Requests are also a parentResource so needed to remove all Organization specific code here.

File: src/client/GraphQLClient/client.test.ts [hunk 17]
```diff
@@ -18,7 +20,7 @@ async function getAccess() {
   const config = context.instance.config;
   sanitizeConfig(config);
   //the installid in the recordings
-  config.installationId = 17214088;
+  config.installationId = 7498286; // TODO: have the recording standardize the installationIds
```
Comment (mknoedel) [inline src/client/GraphQLClient/client.test.ts:23]: - [ ] Fix TODO

File: src/client/GraphQLClient/client.test.ts [hunk 18]
```diff
@@ -4,11 +4,13 @@ import {
...
 } from '@jupiterone/integration-sdk-testing';
 import { setupGithubRecording } from '../../../test/recording';
-import { GitHubGraphQLClient, OrganizationResource } from '.';
+import { GitHubGraphQLClient, GithubResource } from '.';
 import resourceMetadataMap from './resourceMetadataMap';
 import createGitHubAppClient from '../../util/createGitHubAppClient';
 import { IntegrationConfig, sanitizeConfig } from '../../config';
 import { integrationConfig } from '../../../test/config';
+import { Commit, Label, PullRequest, Review } from './types';
+jest.setTimeout(20000);
```
Comment (mknoedel) [inline src/client/GraphQLClient/client.test.ts:13]: - [ ] Remove timeout

File: src/client/GraphQLClient/client.ts [hunk 19]
```diff
@@ -39,85 +51,198 @@ export class GitHubGraphQLClient {
     this.logger = logger;
   }
```
Comment (mknoedel) [inline src/client/GraphQLClient/client.ts:53]: - [ ] Document

File: src/client/GraphQLClient/client.ts [hunk 20]
```diff
@@ -39,85 +51,198 @@ export class GitHubGraphQLClient {
...
+      const response = await query({
+        ...extraQueryParams,
         ...queryCursors,
       });
 
       const rateLimit = response.rateLimit;
       rateLimitConsumed += rateLimit.cost;
 
-      const organizationData = response.organization;
+      const pathToData = this.resourceMetadataMap[baseResource]
```
Comment (mknoedel) [inline src/client/GraphQLClient/client.ts:191]: The path to the data that we actually care about is different between the organization and the pull request.

File: src/client/GraphQLClient/resourceMetadataMap.ts [hunk 21]
```diff
@@ -9,134 +9,215 @@ export default function (
...
+      children: [
+        GithubResource.OrganizationMembers,
+        GithubResource.Teams,
+        GithubResource.TeamMembers,
+        GithubResource.Repositories,
+      ],
     },
-    [OrganizationResource.Members]: {
-      graphRequestVariable: '$members: String',
-      graphProperty: 'membersWithRole',
```
Comment (mknoedel) [inline src/client/GraphQLClient/resourceMetadataMap.ts:23]: got rid of graph property as it was a duplicate and reused it for both the cursor and query names

File: src/client/GraphQLClient/response.ts [hunk 22]
```diff
@@ -72,19 +72,21 @@ export function mapResponseResourcesForQuery(
...
-  const resources: ResourceMap<any> = {};
-  const cursors: ResourceMap<CursorHierarchy> = {};
-
-  extractSelectedResourceFromData(
+  base: GithubResource,
+): {
+  resources: ResourceMap<any>;
+  cursors: ResourceMap<CursorHierarchy>;
+} {
+  const { resources, cursors } = extractSelectedResourceFromData(
```
Comment (mknoedel) [inline src/client/GraphQLClient/response.ts:83]: made it so resources and cursors were returned instead of modified in the `extractSelectedResourceFromData` method.

File: src/client/GraphQLClient/response.ts [hunk 23]
```diff
@@ -98,25 +100,29 @@ function extractSelectedResourceFromData(
...
   const node: { [key: string]: any } = { ...edge };
 
   for (const [key, value] of Object.entries(data)) {
     if (value && typeof value === 'object' && value.edges) {
       const nestedResource = Object.keys(resourceMetadataMap).find(
         (resourceKey) => {
           return (
-            resourceMetadataMap[resourceKey].graphProperty === key &&
+            (resourceMetadataMap[resourceKey].alternateGraphProperty ??
+              resourceKey) === key &&
```
Comment (mknoedel) [inline src/client/GraphQLClient/response.ts:119]: default to the resource key and implemented `alternateGraphProperty` for teamRepos because of its parent flow. The parent flow can be removed in the future thus allowing `alternateGraphProperty` to be removed as well.

File: src/client/GraphQLClient/types/state.ts [hunk 24]
```diff
@@ -14,17 +14,17 @@ export interface NestedQueryFactory {
  */
 export interface ResourceMetadata {
   /*
```
Comment (mknoedel) [inline src/client/GraphQLClient/types/state.ts:16]: TODO: update the documentation on this whole page.

File: src/client/OrganizationAccountClient.ts [hunk 25]
```diff
@@ -405,177 +411,20 @@ export default class OrganizationAccountClient {
     }
   }
 
-  async getPullRequestEntity(
```
Comment (mknoedel) [inline src/client/OrganizationAccountClient.ts:408]: This method currently isn't being used as single entity ingestion isn't supported yet. This will need to be reimplemented using the v4 pull request api later.

File: src/client/OrganizationAccountClient.ts [hunk 26]
```diff
@@ -405,177 +411,20 @@ export default class OrganizationAccountClient {
...
-
-      this.v3RateLimitConsumed++;
-      return commits;
-    } catch (err) {
-      this.logger.info({ err, listOptions }, 'pulls.listCommits failed');
-
-      return [];
+      return { rateLimitConsumed: 0 };
     }
+    // TODO: add sort and update
```
Comment (mknoedel) [inline src/client/OrganizationAccountClient.ts:421]: TODO: sort and update

File: src/client/OrganizationAccountClientErrors.test.ts [hunk 27]
```diff
@@ -69,107 +41,3 @@ describe('handles undefined resources from graphql', () => {
     await expect(client.getMembers()).resolves.toEqual([]);
   });
 });
```
Comment (mknoedel) [inline src/client/OrganizationAccountClientErrors.test.ts:43]: The graphQL api does not handle its own errors right now. This will need to be added in another PR.

File: src/sync/utils/toPullRequestEntityOld.ts [hunk 28]
```diff
@@ -0,0 +1,144 @@
+import {
```
Comment (mknoedel) [inline src/sync/utils/toPullRequestEntityOld.ts:1]: Keeping this around for tests showing that there is no change between the converters. Will remove this and the tests after this PR is merged in

File: src/steps/index.test.ts [hunk 29]
```diff
@@ -29,20 +28,10 @@ test('should collect data', async () => {
...
-          query: false,
-        },
-      },
-    },
   });
 
   sanitizeConfig(integrationConfig);
-  integrationConfig.installationId = 17214088; //this is the id the recordings are under
-  integrationConfig.githubAppDefaultLogin = 'Kei-Institute'; //this is the login for the recordings
+  integrationConfig.installationId = 19232829; //this is the id the recordings are under
```
Comment (mknoedel) [inline src/steps/index.test.ts:34]: - [ ] Make the installationIds standardized in the recordings.

File: src/sync/converters.ts [hunk 30]
```diff
@@ -396,3 +263,179 @@ export function createRepoAllowsUserRelationship(
     pullPermission: true, //always true if there is a relationship
   };
 }
+
+export function toPullRequestEntity(
```
Comment (mknoedel) [inline src/sync/converters.ts:267]: We can move the commit analysis logic into the converter now that all the data is always available.

File: src/sync/converters.ts [hunk 31]
```diff
@@ -396,3 +263,179 @@ export function createRepoAllowsUserRelationship(
...
+        _type: GITHUB_PR_ENTITY_TYPE,
+        _class: [GITHUB_PR_ENTITY_CLASS],
+        _key: `${pullRequest.baseRepository.nameWithOwner}/pull-requests/${pullRequest.number}`,
+        displayName: `${pullRequest.baseRepository.name}/${pullRequest.number}`,
+        accountLogin: pullRequest.baseRepository.owner?.login
+          ? pullRequest.baseRepository.owner.login
+          : '',
+        repository: pullRequest.baseRepository.name,
+        // The number is NOT the id of the Pull Request. Hopefully no one gets bit from that later
+        id: pullRequest.number ? String(pullRequest.number) : '',
```
Comment (mknoedel) [inline src/sync/converters.ts:307]: Should I just change this to the correct value and be done with it?

File: src/sync/converters.ts [hunk 32]
```diff
@@ -396,3 +263,179 @@ export function createRepoAllowsUserRelationship(
...
+          : undefined,
+
+        commits: commitHashes,
+        commitMessages: commits?.map((c) => c.message),
+        commitsApproved: approvedCommitHashes,
+        commitsNotApproved,
+        commitsByUnknownAuthor: commitsByUnknownAuthor?.map((c) => c.oid),
+        validated: commitsByUnknownAuthor
+          ? commitsByUnknownAuthor.length === 0
+          : undefined,
```
Comment (mknoedel) [inline src/sync/converters.ts:336]: I guess `validated` means that all commits are by either members of the organization, or known collaborators. Is that what we want here?

File: src/sync/converters.ts [hunk 33]
```diff
@@ -396,3 +263,179 @@ export function createRepoAllowsUserRelationship(
...
+/**
+ * PULL REQUEST HELPER FUNCTIONS
+ */
+export interface Approval {
+  commit: string;
+  approverUsernames: string[];
+}
+
+function isApprovalReview(review: Review) {
+  return ['APPROVED', 'DISMISSED'].includes(review.state); // Not sure why dismissed is an approved state tbh
```
Comment (mknoedel) [inline src/sync/converters.ts:378]: This was ported, should I change it to just include "APPROVED"?

File: src/client.ts (hunk 12 above)
Comment (mknoedel) [reply src/client.ts:209]: https://github.com/JupiterOne/graph-github/issues/43

File: src/client/OrganizationAccountClient.ts (hunk 26 above)
Comment (mknoedel) [reply src/client/OrganizationAccountClient.ts:421]: This will be done in the next PR.

Comment (mknoedel) [reply src/client/OrganizationAccountClient.ts:421]: https://github.com/JupiterOne/graph-github/issues/44

File: src/client/OrganizationAccountClientErrors.test.ts (hunk 27 above)
Comment (mknoedel) [reply src/client/OrganizationAccountClientErrors.test.ts:43]: https://github.com/JupiterOne/graph-github/issues/45

File: src/steps/index.test.ts (hunk 29 above)
Comment (mknoedel) [reply src/steps/index.test.ts:34]: This will be done in another Pull Request.

Comment (mknoedel) [reply src/steps/index.test.ts:34]: https://github.com/JupiterOne/graph-github/issues/46

File: src/sync/converters.ts (hunk 33 above)
Comment (aiwilliams) [reply src/sync/converters.ts]: Hrm, good question. 🤔  If the function were named `isApprovedReview` I would think it good match only on `APPROVED`. I wonder if "approval review" means, was the review more than just a set of comments? I think we'd have to ask @fomentia if he recalls the reason for this behavior, and then add comments to the function if we keep it as-is.

File: src/client/GraphQLClient/client.test.ts [hunk 34]
```diff
@@ -31,16 +33,118 @@ async function getAccess() {
...
+
+  test('pullRequest with inner pagination', async () => {
+    p = setupGithubRecording({
+      directory: __dirname,
+      name: 'GitHubGraphQLClient.fetchPullRequests.innerPagination',
+    });
+    const client = await getClient();
+
+    const query =
+      'is:pr repo:JupiterOne/graph-whitehat is:closed updated:<=2019-04-01 ';
```
Comment (aiwilliams) [inline src/client/GraphQLClient/client.test.ts:116]: Does this mean we can now work from the last execution time instead of simply getting the last 100?

File: src/client/GraphQLClient/fragments.ts [hunk 35]
```diff
@@ -44,4 +44,94 @@ export default {
...
+    closed
+    closedAt
+    # comments(first: 100) {  # Might be nice to know who is commenting
+    # commits(first: 100) {  # Handling in the query builder
+    createdAt
+    deletions
+    editor {
+      ...userFields
+    }
+    # files(first: 100) {  # Someday it would be great to connect PRs to files that create infrastructure
```
Comment (aiwilliams) [inline src/client/GraphQLClient/fragments.ts]: Do you mean, as an array of values on the PR entity?

File: src/steps/__snapshots__/index.test.ts.snap [hunk 36]
```diff
@@ -621,94 +730,97 @@ Object {
       "_class": Array [
         "PR",
       ],
-      "_key": "Kei-Institute/Test-repo/pull-requests/2",
+      "_key": "undefined/pull-requests/2",
```
Comment (aiwilliams) [inline src/steps/__snapshots__/index.test.ts.snap:733]: Is `undefined` expected here?

**Review by aiwilliams — DISMISSED** [review]
There be tons of code here, and I've never worked on it before, but it looks like there are some great changes! It seems this paves the way for requesting PRs that have changed since the last execution time, which would be 🔥 .

**Review by austinkelleher — COMMENTED** [review]
Looks good overall. Nice changes! I'd like to get this comment addressed before approving: https://github.com/JupiterOne/graph-github/pull/42#discussion_r703958775

**Comment by austinkelleher** [comment]
@mknoedel Thanks for adding comments throughout the PR to guide the review

File: src/sync/converters.ts (hunk 31 above)
Comment (mknoedel) [reply src/sync/converters.ts:308]: I agree with that. I also think that data migrations like that are risky and costly so we should probably steer clear from doing this unless it is necessary.

File: src/sync/converters.ts (hunk 32 above)
Comment (mknoedel) [reply src/sync/converters.ts:337]: Cool, known authors do include external collaborators so that what validated means now. No changes needed.

File: src/client/GraphQLClient/client.test.ts (hunk 34 above)
Comment (mknoedel) [reply src/client/GraphQLClient/client.test.ts:116]: It sure does! That was one of the main reasons to make this change.

File: src/client/GraphQLClient/fragments.ts (hunk 35 above)
Comment (mknoedel) [reply src/client/GraphQLClient/fragments.ts]: Yeah, we can add files as a sub-object on the PRs which we can then use to make file entities in the future.

File: src/steps/__snapshots__/index.test.ts.snap (hunk 36 above)
Comment (mknoedel) [reply src/steps/__snapshots__/index.test.ts.snap:733]: Good catch, this looks pretty sketchy

- [ ] Figure out why this has a key of undefined.

Comment (mknoedel) [reply src/steps/__snapshots__/index.test.ts.snap]: Turns out the graphQL fragment was on `User` instead of `RepositoryOwner` even though they contain the same values. 🙃

File: src/client/GraphQLClient/client.ts [hunk 37]
```diff
@@ -75,12 +78,23 @@ export class GitHubGraphQLClient {
...
+        { pullRequestQueryString, query, queryCursors },
+        'Fetching batch of pull requests from GraphQL',
+      );
+      const pullRequestResponse = await this.retryGraphQL(async () => {
+        return await queryPullRequests({
+          query,
+          ...queryCursors,
+        });
       });
+      pullRequestsQueried += 50; // This is a temporary counter as a stopgap before we get rolling ingestion working for this integration
```
Comment (aiwilliams) [inline src/client/GraphQLClient/client.ts:91]: Should this be a reference to `limit` (100)?

File: src/client/GraphQLClient/client.ts [hunk 38]
```diff
@@ -250,4 +276,29 @@ export class GitHubGraphQLClient {
...
+  private async retryGraphQL(query: () => Promise<any>) {
+    const { logger } = this;
+    return await retry(query, {
+      maxAttempts: 10,
+      delay: 60_000,
+      handleError(error: any, attemptContext: AttemptContext) {
+        // Github has "Secondary Rate Limits" to prevent us from making these costly to Github graphQL calls.
+        // From what I can tell, there is no way around it outside of waiting for "a few minutes" when we get
+        // one of these errors. I guess checking every minute for 10 minutes seems reasonable.
+        // https://docs.github.com/en/rest/overview/resources-in-the-rest-api#secondary-rate-limits
```
Comment (aiwilliams) [inline src/client/GraphQLClient/client.ts:289]: They state in that documentation that we should be using Webhooks. 😆  I suppose this doc, which references REST API concerns, applies to the GraphQL API as well?

**Review by aiwilliams — APPROVED** [review]
In your opening comments, the two images for 'old' and 'new', there is a substantial difference in the upload counts. Why is that?

File: src/client/GraphQLClient/client.ts (hunk 37 above)
Comment (mknoedel) [reply src/client/GraphQLClient/client.ts:91]: It shouldn't reference the limit, but it should likely reference the query value in the resourceMetadataMap. Though, because of the way it is coded, it's hard to parse it out correctly. I decided that because this is a stopgap, doing this is fine for now as it will be ripped out once we add rolling ingestion.
